profiler/profiler.go

```diff
@@ -69,6 +71,42 @@ type profiler struct {
 	wg         sync.WaitGroup                    // wg waits for all goroutines to exit when stopping.
 	met        *metrics                          // metric collector state
 	prev       map[ProfileType]*pprofile.Profile // previous collection results for delta profiling
+
+	testHooks testHooks
+}
+
+// testHooks are functions that are replaced during testing which would normally
+// depend on accessing runtime state that is not needed/available for the test
+type testHooks struct {
+	startCPUProfile func(w io.Writer) error
+	stopCPUProfile  func()
```

Comment:
I think it would be doable. Most of the `lookupProfile` uses are just to put _something_ in the profile output and check that it comes through (or nothing in the case of `TestRunProfile/goroutineswaitLimit`). For that use-case, `TestAllUploaded`-style tests can capture that while running the real profiles.

The `TestRunProfile/goroutineswait` test could be replaced with a standalone test of `goroutineDebug2ToPprof` since I think the test is really looking for whether a goroutine debug stack trace gets converted to a pprof correctly.

The other slightly complicated use of replacing `lookupProfile` is in `TestRunProfile/delta`. But I think that test is basically a duplicate of the delta profile tests in `profiler/internal/pprofutils` so practically speaking we don't lose anything by just deleting the test completely. We could have `TestAllUploaded` variants that check whether delta or non-delta profiles get uploaded, at least.